File: Category/math/qpow.cpp
// 乘法快速幂，带取模
template <typename T = uint64_t>
T qpow(T a, T b, int p) {
    T res = 1;
    a = (a % p + p) % p;
    for (; b; b >>= 1, a = a * a % p) {
        if (b & 1) {
            res = a * res % p;
        }
    }
    return res;
}

// 乘法快速幂，未取模
template <typename T = uint64_t>
T qpow(T a, T b) {
    T res = 1;
    for (; b; b >>= 1, a *= a) {
        if (b & 1) {
            res = res * a;
        }
    }
    return res;
}

// 矩阵快速幂（方阵）
template <typename T = uint64_t>
vector<vector<T>> multiply(vector<vector<T>>& a, vector<vector<T>>& b) { // 矩阵乘法
    int m = a.size(), n = b[0].size(), k = a[0].size();
    vector<vector<T>> res(m, vector<T>(n));
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            for (int c = 0; c < k; ++c) {
                res[i][j] += a[i][c] * b[c][j];
                // res[i][j] %= MOD;
            }
        }
    }
    return res;
}
template <typename T = uint64_t>
vector<vector<T>> qpow(vector<vector<T>> a, T b) {
    int n = a.size();
    vector<vector<T>> ans(n, vector<T>(n));
    for (int i = 0; i < n; ++i) {
        ans[i][i] = 1;
    }
    for (; b; b >>= 1) {
        if (b & 1) {
            ans = multiply(ans, a);
        }
        a = multiply(a, a);
    }
    return ans;
}

File: Category/math/qpow.py
def qpow(a: int, b: int, p: int):
    ans = 1
    while b:
        if b & 1:
            ans = ans * a % p
        a = a * a % p
        b >>= 1
    return ans


def qpow(a: int, b: int):
    ans = 1
    while b:
        if b & 1:
            ans *= a
        a *= a
        b >>= 1
    return ans


# a @ b，其中 @ 是矩阵乘法
def multiply(a: List[List[int]], b: List[List[int]]) -> List[List[int]]:
    return [
        [sum(x * y for x, y in zip(row, col)) % MOD for col in zip(*b)] for row in a
    ]


# a^n
def qpow(a: List[List[int]], b: int) -> List[List[int]]:
    n = len(a)
    res = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
    while b:
        if b & 1:
            res = multiply(a, res)
        a = multiply(a, a)
        b >>= 1
    return res


# a^n @ f0
def qpow_mul(a: List[List[int]], b: int, f: List[List[int]]) -> List[List[int]]:
    res = f
    while b:
        if b & 1:
            res = multiply(a, res)
        a = multiply(a, a)
        b >>= 1
    return res
